#!/bin/sh -e
## DP: Pluto changes to icewm

if [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch}"

case "$1" in
       -patch) patch $patch_opts -p1 < $0;;
       -unpatch) patch $patch_opts -p1 -R < $0;;
        *)
                echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
                exit 1;;
esac

exit 0

@DPATCH@
diff -Naur icewm-1.2.26.orig/src/wmcontainer.cc icewm-1.2.26/src/wmcontainer.cc
--- icewm-1.2.26.orig/src/wmcontainer.cc	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/wmcontainer.cc	2006-07-21 16:41:02.000000000 +0000
@@ -14,8 +14,8 @@
 
 #include <stdio.h>
 
-YClientContainer::YClientContainer(YWindow *parent, YFrameWindow *frame)
-:YWindow(parent)
+YClientContainer::YClientContainer(YWindow *parent, YFrameWindow *frame, Visual *visual)
+    :YWindow(parent, 0, visual)
 {
     fFrame = frame;
     fHaveGrab = false;
diff -Naur icewm-1.2.26.orig/src/wmcontainer.h icewm-1.2.26/src/wmcontainer.h
--- icewm-1.2.26.orig/src/wmcontainer.h	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/wmcontainer.h	2006-07-21 16:41:02.000000000 +0000
@@ -7,7 +7,7 @@
 
 class YClientContainer: public YWindow {
 public:
-    YClientContainer(YWindow *parent, YFrameWindow *frame);
+    YClientContainer(YWindow *parent, YFrameWindow *frame, Visual *visual);
     virtual ~YClientContainer();
 
     virtual void handleButton(const XButtonEvent &button);
diff -Naur icewm-1.2.26.orig/src/wmframe.cc icewm-1.2.26/src/wmframe.cc
--- icewm-1.2.26.orig/src/wmframe.cc	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/wmframe.cc	2006-07-21 16:41:02.000000000 +0000
@@ -48,7 +48,7 @@
     return false;
 }
 
-YFrameWindow::YFrameWindow(YWindow *parent): YWindow(parent) {
+YFrameWindow::YFrameWindow(YWindow *parent, Visual *visual) : YWindow(parent, 0, visual) {
     if (activeBorderBg == 0)
         activeBorderBg = new YColor(clrActiveBorder);
     if (inactiveBorderBg == 0)
@@ -126,7 +126,7 @@
 
     createPointerWindows();
 
-    fClientContainer = new YClientContainer(this, this);
+    fClientContainer = new YClientContainer(this, this, visual);
     fClientContainer->show();
 
     fTitleBar = new YFrameTitleBar(this, this);
diff -Naur icewm-1.2.26.orig/src/wmframe.h icewm-1.2.26/src/wmframe.h
--- icewm-1.2.26.orig/src/wmframe.h	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/wmframe.h	2006-07-21 16:41:02.000000000 +0000
@@ -20,7 +20,7 @@
 
 class YFrameWindow: public YWindow, public YActionListener, public YTimerListener, public YPopDownListener, public YMsgBoxListener, public ClientData {
 public:
-    YFrameWindow(YWindow *parent);
+    YFrameWindow(YWindow *parent, Visual *visual = 0);
     virtual ~YFrameWindow();
 
     void doManage(YFrameClient *client);
diff -Naur icewm-1.2.26.orig/src/wmmgr.cc icewm-1.2.26/src/wmmgr.cc
--- icewm-1.2.26.orig/src/wmmgr.cc	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/wmmgr.cc	2006-07-21 16:41:02.000000000 +0000
@@ -1374,7 +1374,7 @@
 
     manager->updateFullscreenLayerEnable(false);
 
-    frame = new YFrameWindow(0);
+    frame = new YFrameWindow(0, client->visual());
     if (frame == 0) {
         delete client;
         goto end;
diff -Naur icewm-1.2.26.orig/src/ywindow.cc icewm-1.2.26/src/ywindow.cc
--- icewm-1.2.26.orig/src/ywindow.cc	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/ywindow.cc	2006-07-21 16:49:19.000000000 +0000
@@ -18,6 +18,8 @@
 #include "ytimer.h"
 #include "ypopup.h"
 
+#include <X11/extensions/Xrender.h>
+
 /******************************************************************************/
 /******************************************************************************/
 
@@ -114,7 +116,7 @@
 /******************************************************************************/
 /******************************************************************************/
 
-YWindow::YWindow(YWindow *parent, Window win):
+YWindow::YWindow(YWindow *parent, Window win, Visual *visual):
     fParentWindow(parent),
     fNextWindow(0), fPrevWindow(0), fFirstWindow(0), fLastWindow(0),
     fFocusedWindow(0),
@@ -122,6 +124,7 @@
     fHandle(win), flags(0), fStyle(0), fX(0), fY(0), fWidth(1), fHeight(1),
     fPointer(), unmapCount(0), 
     fGraphics(0),
+    fVisual(visual),
     fEventMask(KeyPressMask|KeyReleaseMask|FocusChangeMask|
                LeaveWindowMask|EnterWindowMask),
     fWinGravity(NorthWestGravity), fBitGravity(ForgetGravity),
@@ -301,7 +304,7 @@
                                 0,
                                 CopyFromParent,
                                 (fStyle & wsInputOnly) ? InputOnly : InputOutput,
-                                CopyFromParent,
+                                (fVisual) ? fVisual : CopyFromParent,
                                 attrmask,
                                 &attributes);
 
@@ -311,6 +314,9 @@
 
         if ((flags & wfVisible) && !(flags & wfNullSize))
             XMapWindow(xapp->display(), fHandle);
+
+        // Zero visual so it is not used for system menu
+        fVisual = 0;
     } else {
         XWindowAttributes attributes;
 
@@ -352,6 +358,22 @@
         }
 
         XSelectInput(xapp->display(), fHandle, fEventMask);
+
+        // Get the visual if we have a window with alpha transparency
+        XWindowAttributes attr;
+        Status err = XGetWindowAttributes(xapp->display(), fHandle, &attr);
+        bool ok = (BadDrawable != err) && (BadWindow != err) && attr.visual;
+
+        XRenderPictFormat *format = NULL;
+        if (ok)
+            format = XRenderFindVisualFormat(xapp->display(), attr.visual);
+
+        if (format && format->type == PictTypeDirect &&
+            format->direct.alphaMask)
+        {
+            //printf("YWindow(0x%X) adopted alpha window\n", this);
+            fVisual = attr.visual;
+        }
     }
     XSaveContext(xapp->display(), fHandle, windowContext, (XPointer)this);
     flags |= wfCreated;
diff -Naur icewm-1.2.26.orig/src/ywindow.h icewm-1.2.26/src/ywindow.h
--- icewm-1.2.26.orig/src/ywindow.h	2006-04-16 16:12:31.000000000 +0000
+++ icewm-1.2.26/src/ywindow.h	2006-07-21 16:41:02.000000000 +0000
@@ -24,7 +24,7 @@
 
 class YWindow {
 public:
-    YWindow(YWindow *aParent = 0, Window win = 0);
+    YWindow(YWindow *aParent = 0, Window win = 0, Visual *visual = 0);
     virtual ~YWindow();
 
     void setStyle(unsigned long aStyle);
@@ -202,6 +202,8 @@
     bool hasPopup();
     void setDoubleBuffer(bool doubleBuffer);
 
+    Visual *visual() { return fVisual; }
+
 private:
     typedef enum {
         wfVisible   = 1 << 0,
@@ -234,6 +236,7 @@
     YCursor fPointer;
     int unmapCount;
     Graphics *fGraphics;
+    Visual *fVisual;
     long fEventMask;
     int fWinGravity, fBitGravity;
 
